### PR TITLE
Use Limelight camera to calculate location to travel to

### PR DESCRIPTION
**Pull request by nikwalia**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -119,6 +119,7 @@ lazy val frc = crossProject(JVMPlatform, NativePlatform).crossType(CrossType.Ful
   libraryDependencies += "edu.wpi.first" % "ntcore" % wpiVersion,
   libraryDependencies += "com.ctre" % "phoenix" % ctreVersion
 ).nativeSettings(
+  resolvers += "Funky-Repo" at "http://lynbrookrobotics.com/repo",
```
Comment (shadaj) [inline build.sbt]: You should be able to reset this to master and everything should work.

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala [hunk 2]
```diff
@@ -385,4 +385,34 @@ trait UnicycleCoreTasks {
     }
   }
 
+  class DriveToTarget(drivetrainComponent: Drivetrain, distanceToTarget: Stream[Option[Length]], angleToTarget: Stream[Angle])
+                    (implicit val drivetrainHardware: DrivetrainHardware,
```
Comment (shadaj) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala]: Nit: parens should be aligned

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala [hunk 3]
```diff
@@ -385,4 +385,34 @@ trait UnicycleCoreTasks {
...
+                      implicit val props: Signal[DrivetrainProperties]) extends FiniteTask {
+
```
Comment (shadaj) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala]: Nit: extra line

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala [hunk 4]
```diff
@@ -385,4 +385,34 @@ trait UnicycleCoreTasks {
...
+          if (!p.exists(_ >= Feet(2))) {
+            finished()
+          }
+      )))
+    }
+
+    override def onEnd(): Unit = {
+      drivetrainComponent.resetToDefault()
+    }
+
```
Comment (shadaj) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala]: Nit: extra line

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala [hunk 5]
```diff
@@ -385,4 +385,34 @@ trait UnicycleCoreTasks {
...
+  }
+
```
Comment (shadaj) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala]: Nit: extra line

File: vision/src/main/scala/com/lynbrookrobotics/potassium/vision/VisionTargetTracking.scala [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+      p.map { percentArea =>
+        camProps.get.distanceConstant / math.sqrt(percentArea.toPercent)
+      }
+    )
+  }
+
+  def angleToTarget(xOffset: Stream[Angle], camProps: Signal[VisionProperties]): Stream[Angle] = {
+    xOffset.map(-_ - camProps.get.cameraHorizontalOffset)
+  }
+}
```
Comment (shadaj) [inline vision/src/main/scala/com/lynbrookrobotics/potassium/vision/VisionTargetTracking.scala]: Nit: file should end with new line

File: vision/src/test/scala/com/lynbrookrobotics/potassium/vision/TargetTrackingTest.scala [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.lynbrookrobotics.potassium.Signal
+import com.lynbrookrobotics.potassium.streams._
+import org.scalatest.FunSuite
+import squants.{Angle, Dimensionless, Length, Percent}
+import squants.space.{Degrees, Feet}
+import squants.time.Milliseconds
+
+class TargetTrackingTest extends FunSuite {
+
```
Comment (shadaj) [inline vision/src/test/scala/com/lynbrookrobotics/potassium/vision/TargetTrackingTest.scala]: Nit: extra line

File: vision/src/test/scala/com/lynbrookrobotics/potassium/vision/TargetTrackingTest.scala [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  test("cube with measured percent area 15.6 results in distance 2.75 feet away") {
```
Comment (shadaj) [inline vision/src/test/scala/com/lynbrookrobotics/potassium/vision/TargetTrackingTest.scala]: Nit: test should start with capital letter

File: vision/src/test/scala/com/lynbrookrobotics/potassium/vision/TargetTrackingTest.scala [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    var lastTargetPosition: Option[Length] = Option(Feet(0))
+
+    distanceToTarget.foreach(lastTargetPosition = _)
+
+    pubPercentArea(Some(Percent(15.6)))
+
+    assert(targetDistance.get - lastTargetPosition.get < Feet(0.1))
+  }
+
+  test("angle to target measured as 5 degrees for input of 10 degrees and offset of -5 degrees") {
```
Comment (shadaj) [inline vision/src/test/scala/com/lynbrookrobotics/potassium/vision/TargetTrackingTest.scala]: Nit: test should start with capital letter

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala [hunk 10]
```diff
@@ -385,4 +385,34 @@ trait UnicycleCoreTasks {
     }
   }
 
+  class DriveToTarget(drivetrainComponent: Drivetrain, distanceToTarget: Stream[Option[Length]], angleToTarget: Stream[Angle])
```
Comment (shadaj) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala]: This task definitely needs unit tests.

File: vision/src/main/scala/com/lynbrookrobotics/potassium/vision/limelight/LimelightNetwork.scala [hunk 11]
```diff
@@ -0,0 +1,32 @@
+package com.lynbrookrobotics.potassium.vision.limelight
+
+import com.lynbrookrobotics.potassium.clock.Clock
+import com.lynbrookrobotics.potassium.streams._
+import edu.wpi.first.networktables.{NetworkTable, NetworkTableInstance}
+import squants.{Dimensionless, Percent}
+import squants.space.{Angle, Degrees}
+import squants.time.Milliseconds
+
+class LimelightNetwork(val table: NetworkTable)(implicit clock: Clock) {
```
Comment (shadaj) [inline vision/src/main/scala/com/lynbrookrobotics/potassium/vision/limelight/LimelightNetwork.scala]: Can you right unit tests for this with a Mockito-mocked `NetworkTable`?

**Review by shadaj — CHANGES_REQUESTED** [review]

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala [hunk 12]
```diff
@@ -385,4 +385,38 @@ trait UnicycleCoreTasks {
...
+                                       distanceToTarget: Stream[Option[Length]],
+                                       angleToTarget: Stream[Angle],
+                                       forwardVelocity: Dimensionless,
+                                       maxTurnVelocity: Dimensionless,
+                                       minDistance: Length)
+                                      (implicit drivetrainHardware: DrivetrainHardware,
+                                       props: Signal[DrivetrainProperties]) extends FiniteTask {
+
+    override def onStart(): Unit = {
+      val absoluteTargetAngle = drivetrainHardware.turnPosition.zipAsync(angleToTarget).map{t =>
```
Comment (shadaj) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala]: Nit: space before curly

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala [hunk 13]
```diff
@@ -385,4 +385,38 @@ trait UnicycleCoreTasks {
...
+        t._1 + t._2
+      }
+
+      val turnController = turnPositionControl(absoluteTargetAngle)._1
+
+      val out = childVelocityControl(
+        turnController.map{p =>
```
Comment (shadaj) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala]: Nit: space before curly

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala [hunk 14]
```diff
@@ -385,4 +385,38 @@ trait UnicycleCoreTasks {
...
+          UnicycleSignal(forwardVelocity, p.turn min maxTurnVelocity max -maxTurnVelocity)
+        }
+      )
+
+      drivetrainComponent.setController(out.withCheck(_ =>
+        distanceToTarget.foreach(p =>
```
Comment (shadaj) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala]: Whoa! What is going on here? In the current implementation, you are adding a new listener to `distanceToTarget` on every tick of `out`. You should be using `withCheckZipped` instead.

File: vision/src/main/scala/com/lynbrookrobotics/potassium/vision/VisionTargetTracking.scala (hunk 6 above)
Comment (shadaj) [reply vision/src/main/scala/com/lynbrookrobotics/potassium/vision/VisionTargetTracking.scala]: This is still an issue.

File: vision/src/main/scala/com/lynbrookrobotics/potassium/vision/VisionProperties.scala [hunk 15]
```diff
@@ -0,0 +1,6 @@
+package com.lynbrookrobotics.potassium.vision
+
+import squants.Angle
+import squants.space.Length
+
+case class VisionProperties(cameraHorizontalOffset: Angle, distanceConstant: Length)
```
Comment (shadaj) [inline vision/src/main/scala/com/lynbrookrobotics/potassium/vision/VisionProperties.scala]: This class can probably just be pulled into `VisionTargetTracking`.

File: vision/src/test/scala/com/lynbrookrobotics/potassium/vision/TargetTrackingTest.scala [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    val (percentArea, pubPercentArea) = Stream.manual[Option[Dimensionless]]
+    val camProps: Signal[VisionProperties] = Signal.constant(VisionProperties(Degrees(0), Feet(10.8645)))
+    val targetDistance: Option[Length] = Some(Feet(2.75))
+    val distanceToTarget = VisionTargetTracking.distanceToTarget(percentArea, camProps)
+    var lastTargetPosition: Option[Length] = Option(Feet(0))
+
+    distanceToTarget.foreach(lastTargetPosition = _)
+    pubPercentArea(Some(Percent(15.6)))
+
+    assert(targetDistance.get - lastTargetPosition.get < Feet(0.1))
```
Comment (shadaj) [inline vision/src/test/scala/com/lynbrookrobotics/potassium/vision/TargetTrackingTest.scala]: You should be setting an `implicit val tolerance = Feet(0.1)` and then doing `lastTargetPosition.get ~= Feet(2.75)` to make this test more clear.

File: vision/src/test/scala/com/lynbrookrobotics/potassium/vision/TargetTrackingTest.scala [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    val (angle, pubAngle) = Stream.manual[Angle]
+    val camProps: Signal[VisionProperties] = Signal.constant(VisionProperties(Degrees(-5), Feet(10.8645)))
+    val targetAngle: Angle = Degrees(-5)
+    val angleToTarget = VisionTargetTracking.angleToTarget(angle, camProps)
+    var lastAngle = Degrees(0)
+
+    angleToTarget.foreach(lastAngle = _)
+    pubAngle(Degrees(10))
+
+    assert(targetAngle - lastAngle < Degrees(1))
```
Comment (shadaj) [inline vision/src/test/scala/com/lynbrookrobotics/potassium/vision/TargetTrackingTest.scala]: Same as above, its harder to figure out what this is doing when I have to search the code for the `targetAngle` value. Would be nicer if this was a direct comparison between lastAngle and what you expect it to be.

**Review by shadaj — CHANGES_REQUESTED** [review]
Couple places that need cleanup but otherwise looks good!

File: commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala [hunk 18]
```diff
@@ -385,4 +385,39 @@ trait UnicycleCoreTasks {
...
+
+      val out = childVelocityControl(
+        turnController.map { p =>
+          UnicycleSignal(forwardVelocity, p.turn min maxTurnVelocity max -maxTurnVelocity)
+        }
+      )
+
+      drivetrainComponent.setController(out.withCheckZipped(distanceToTarget){
+        distanceToTarget => {
+          if (!distanceToTarget.exists(_ > minDistance)) {
```
Comment (shadaj) [inline commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/unicycle/control/UnicycleCoreTasks.scala]: Shouldn't this be a exists < minDistance? Right now this task will finish is no cube is detected.

**Comment by shadaj** [comment]
@nikwalia any updates on this? We really need this PR merged ASAP.

**Review by PhilipAxelrod — APPROVED** [review]
Lgtm!

**Review by PhilipAxelrod — APPROVED** [review]
Lgtm #2 ( because first approval didn't work?)

File: vision/src/main/scala/com/lynbrookrobotics/potassium/vision/VisionTargetTracking.scala [hunk 19]
```diff
@@ -0,0 +1,20 @@
+package com.lynbrookrobotics.potassium.vision
+
+import com.lynbrookrobotics.potassium.Signal
+import com.lynbrookrobotics.potassium.streams.Stream
+import squants.Dimensionless
+import squants.space.{Angle, Length}
+
+class VisionTargetTracking(cameraHorizontalOffset: Signal[Angle], distanceConstant: Signal[Length]) {
+  def distanceToTarget(percentArea: Stream[Option[Dimensionless]]): Stream[Option[Length]] = {
+        percentArea.map ( p =>
```
Comment (shadaj) [inline vision/src/main/scala/com/lynbrookrobotics/potassium/vision/VisionTargetTracking.scala]: Please fix indentation

File: commons/src/test/scala/com/lynbrookrobotics/potassium/commons/drivetrain/UnicycleDriveTaskTest.scala [hunk 20]
```diff
@@ -397,4 +396,92 @@ class UnicycleDriveTaskTest extends FunSuite {
 
     assert(!task.isRunning)
   }
+
+  test("Driving to a given polar target results in target position") {
```
Comment (shadaj) [inline commons/src/test/scala/com/lynbrookrobotics/potassium/commons/drivetrain/UnicycleDriveTaskTest.scala]: This title isn't actually what you're testing.

**Review by shadaj — CHANGES_REQUESTED** [review]

**Review by shadaj — APPROVED** [review]

**Pull request merged**